x/staking/keeper/epoch_exec.go

```diff
@@ -0,0 +1,391 @@
+	}
+
+	completionTime, err := k.Undelegate(ctx, delegatorAddress, addr, shares)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	if msg.Amount.Amount.IsInt64() {
+		defer func() {
+			telemetry.IncrCounter(1, types.ModuleName, "undelegate")
+			telemetry.SetGaugeWithLabels(
+				[]string{"tx", "msg", msg.Type()},
+				float32(msg.Amount.Amount.Int64()),
+				[]metrics.Label{telemetry.NewLabel("denom", msg.Amount.Denom)},
+			)
+		}()
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUnbond,
+			sdk.NewAttribute(types.AttributeKeyValidator, msg.ValidatorAddress),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, msg.Amount.Amount.String()),
+			sdk.NewAttribute(types.AttributeKeyCompletionTime, completionTime.Format(time.RFC3339)),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.DelegatorAddress),
+		),
+	})
+
+	return completionTime, nil
+}
+
+// ExecuteEpoch execute epoch actions
+// If a message fails to execute, the message will be reverted if the tokens were taken from an account,
+// otherwise an error log will be produced.
+// Unfortunately, the user will need to check if all there messages were executed.
+func (k Keeper) ExecuteEpoch(ctx sdk.Context) {
+	// execute all epoch actions
+
+	for iterator := k.epochKeeper.GetEpochActionsIterator(ctx); iterator.Valid(); iterator.Next() {
+		msg := k.GetEpochActionByIterator(iterator)
+		cacheCtx, writeCache := ctx.CacheContext()
+		logger := k.Logger(ctx)
+
+		switch msg := msg.(type) {
+		case *types.MsgCreateValidator:
+			err := k.executeQueuedCreateValidatorMsg(cacheCtx, msg)
+			if err == nil {
+				writeCache()
+			} else if err = k.revertCreateValidatorMsg(ctx, msg); err != nil {
+				// avoid panicking
+				logger.Error("create validator failed to execute", "msg", msg)
+			}
+		case *types.MsgEditValidator:
+			err := k.executeQueuedEditValidatorMsg(cacheCtx, msg)
+			if err == nil {
+				writeCache()
+			} else {
+				// TODO: report somewhere for logging edit not success or panic
+				logger.Error("edit validator failed to execute", "msg", msg)
+				// panic(fmt.Sprintf("not be able to execute, %T", msg))
+			}
+		case *types.MsgDelegate:
+			err := k.executeQueuedDelegationMsg(cacheCtx, msg)
+			if err == nil {
+				writeCache()
+			} else if err = k.revertDelegationMsg(ctx, msg); err != nil {
+				panic(fmt.Sprintf("not be able to execute nor revert, %T", msg))
+			}
+		case *types.MsgBeginRedelegate:
+			_, err := k.executeQueuedBeginRedelegateMsg(cacheCtx, msg)
+			if err == nil {
+				writeCache()
+			} else {
+				logger.Error("edit begin redelegation failed to execute", "msg", msg)
+			}
+		case *types.MsgUndelegate:
+			_, err := k.executeQueuedUndelegateMsg(cacheCtx, msg)
+			if err == nil {
+				writeCache()
+			} else {
+				logger.Error("edit undelegation failed to execute", "msg", msg)
```

Comment:
@marbar3778 does it make sense to emit an error event in addition to logging in ExecuteEpoch? So that external users can know (via logs) why their Msg failed.

Reply:
Will look into this. This would of avoided the issue i was running into.